### PR TITLE
Added auth interface

### PR DESCRIPTION
**Pull request by harryi3t**
This wraps an auth instance and provide helper functions around it
- `set`: allows updating or adding new params to auth
  example:
  ```js
  authInterface = getAuthInterface(reqAuth)
  authInterface.set('foo', 'bar') 
  authInterface.set({'foo': 'bar', 'alpha': 'beta'})
  ```
- `get`: allows gettings param(s)
example:
  ```js
  authInterface = getAuthInterface(reqAuth)
  authInterface.get('foo') // bar 
  authInterface.get(['foo', 'alpha']) // {'foo': 'bar', 'alpha': 'beta'}
  ```

### CODE REVIEW

File: lib/authorizer/auth-interface.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
+var _ = require('lodash'),
+    createAuthInterface;
+
+/**
+* Creates a wrapper around RequestAuth and provides getters and setters helper functions
+*
+* @param {RequestAuth} auth
+* @constructor
```
Comment (kamalaknn) [inline lib/authorizer/auth-interface.js]: @harryi3t this is not a constructor. Also, the return type of this should be `AuthInterface`.

**Comment by harryi3t** [comment]
Before this can be merged, need to update SDK's Property to support `system` property

File: lib/authorizer/auth-interface.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    createAuthInterface;
+
+/**
+ * Creates a wrapper around RequestAuth and provides getters and setters helper functions
+ *
+ * @param {RequestAuth} auth
+ * @return {AuthInterface}
+ */
+createAuthInterface = function (auth) {
+    if (!(auth && auth.parameters && auth.parameters())) {
```
Comment (shamasis) [inline lib/authorizer/auth-interface.js]:
```
if (!_.isObject(auth)) { return; };
```

File: lib/authorizer/auth-interface.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+         * get(['foo', 'alpha']) // {foo: 'bar', 'alpha': 'beta'}
+         */
+        get: function (keys) {
+            var paramVariable;
+            if (_.isString(keys)) {
+                paramVariable = auth.parameters().one(keys);
+                return paramVariable && paramVariable.get();
+            }
+            if (_.isArray(keys)) {
+                return _.reduce(keys, function (paramObject, key) {
```
Comment (shamasis) [inline lib/authorizer/auth-interface.js]: `_.transform`

File: lib/authorizer/auth-interface.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        /**
+         * @param {String|Object} key
+         * @param {*} [value]
+         * @return {AuthInterface}
+         * @example set('foo', 'bar') or set({foo: 'bar', 'alpha': 'beta'})
+         */
+        set: function (key, value) {
+            var modifiedParams = {};
+
+            if (_.isPlainObject(key)) {
```
Comment (shamasis) [inline lib/authorizer/auth-interface.js]: do `_.isObject`

File: lib/authorizer/auth-interface.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                modifiedParams = key;
+            }
+            else if (_.isString(key)) {
+                modifiedParams[key] = value;
+            }
+            else { // invalid input found, no change to auth
+                return this;
```
Comment (shamasis) [inline lib/authorizer/auth-interface.js]: maybe throw up?!

File: lib/authorizer/auth-interface.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            }
+
+            _.forEach(modifiedParams, function (value, key) {
+                var param;
+                // if auth has the param already then update it only when it's a sytem property
+                if ((param = auth.parameters().one(key))) {
```
Comment (shamasis) [inline lib/authorizer/auth-interface.js]: No need to call `.parameters()` in loop. Keep a reference outside forEach

File: test/unit/auth-interface.test.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+        authInterface.set('joker', joker);
+        authInterface.set({'gordon': gordon});
+        expect(authInterface.get('joker')).to.be(joker);
+        expect(authInterface.get('gordon')).to.be(gordon);
+        expect(fakeAuth.parameters().one('joker')).to.have.property('system', true);
+        expect(fakeAuth.parameters().one('gordon')).to.have.property('system', true);
+    });
+
+    it('set with invalid params should return the auth unchanged', function () {
```
Comment (shamasis) [inline test/unit/auth-interface.test.js]: now throw error

File: test/unit/auth-interface.test.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            newUsername = 'bane',
+            newCreds = {user: newUsername}; // only partial update, password & nonce shoudn't change
+
+        authInterface.set(newCreds);
+        expect(authInterface.get('user')).to.be(newUsername);
+        expect(authInterface.get('pass')).to.be(PASS);
+        expect(authInterface.get('nonce')).to.be(NONCE);
+    });
+
+    it('non system parameters should not be able to be updated', function () {
```
Comment (shamasis) [inline test/unit/auth-interface.test.js]: translate this to english please

File: test/unit/auth-interface.test.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+describe('AuthInterface', function () {
+    it('should throw an error on invalid input', function () {
+        expect(createAuthInterface).withArgs({type: 'basic', basic: {}})
+            .to.throwError(/runtime~createAuthInterface: invalid auth/);
+    });
+
+    it('get with single key should return single value', function () {
+        var fakeAuth = new sdk.RequestAuth(fakeAuthObj),
+            authInterface = createAuthInterface(fakeAuth);
```
Comment (shamasis) [inline test/unit/auth-interface.test.js]: data type retention check

**Pull request merged**